File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    // result variable
    let result = 0;
    // create the hashset
    let mySet = new Set();
    
    for(let i=0;i<s.length;i++) {
        
        let currentTemp = s[i];
        if(mySet.has(currentTemp)) {
            // find the index at which the value is at 
            removeElement(mySet, currentTemp)
            // reset the set 
            mySet.add(s[i]);
        }
        else {
            mySet.add(s[i]);
        }
        
        result = Math.max(result, mySet.size);
    }
    return result;    
};

function removeElement(myset, element) {
    for (let item of myset) {
        if(item==element) {
            myset.delete(item);
            break;
        }
        myset.delete(item);
    }
}

File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
    
        
        // result set 
        HashSet<Character> set  = new HashSet<Character>();
        int result = 0;
        
        int a_pointer = 0;
        int b_pointer = 0;
        
        while(b_pointer < s.length()) {
            if(!set.contains(s.charAt(b_pointer))) {
                set.add(s.charAt(b_pointer));
                result = Math.max(set.size(), result);
                b_pointer++;
            } else {
                set.remove(s.charAt(a_pointer));
                a_pointer++;
            }
        }
        
        return result;
       
        
    }
}


/*
1. create a hashset 
2. push values to the hashset - when the push is false - do math.max 
3. reset the hashset



*/